### PR TITLE
Adding typealias and automatically releasing objects

### PR DESCRIPTION
**Pull request by cstjean**
I'd like to add GIF support this week, and I wanted to clean up the code a bit beforehand. I'm not sure I like the typealias changes. It looks noisier, but it's probably for the best?

Also, `save_and_release` never actually released the image, so there was a memory leak, now fixed.

### CODE REVIEW

File: src/QuartzImageIO.jl [hunk 1]
```diff
@@ -246,14 +280,14 @@ function save_(fname, img::Image, image_type)
...
-    
-    save_and_release(cgImage, fname, image_type)
+    with_obj(CGColorSpaceCreateDeviceRGB()) do colspace
+        with_obj(CGBitmapContextCreate(buf, nx, ny, 8, nx*4, colspace,
+                                       kCGImageAlphaPremultipliedLast)) do bmp_context
+            with_CGImage(CGBitmapContextCreateImage(bmp_context)) do cgImage
+                do_save(cgImage, fname, image_type)
+            end
+        end                           
+    end
```
Comment (rsrock) [inline src/QuartzImageIO.jl]: I don't think I like this `with_obj` approach. The block nesting can quickly get out of hand.  I also find the original code much more readable.

**Comment by rsrock** [comment]
Thanks for working on this, sorry that it's taken so long to review.  I like the typealiases, but I'm not so keen on the `with_obj` (see the inline comments).  Could you revert those changes?  If it's easier, you can address the memory leak in a separate PR.

For error handling, I'd like to handle the NULLs and other errors at the call site, like in http://docs.julialang.org/en/latest/manual/calling-c-and-fortran-code/ under the `getenv()` example.   The error handling in the loading code could use some love too, along these lines.

File: src/QuartzImageIO.jl [hunk 2]
```diff
@@ -222,14 +255,15 @@ end
...
 ## Saving Images ###############################################################
 
-function save_and_release(cg_img::Ptr{Void}, # CGImageRef
-                          fname, image_type::AbstractString)
-    out_url = CFURLCreateWithFileSystemPath(fname);
-    out_dest = CGImageDestinationCreateWithURL(out_url, image_type, 1);
-    CGImageDestinationAddImage(out_dest, cg_img);
-    CGImageDestinationFinalize(out_dest);
-    CFRelease(out_dest)
-    CFRelease(out_url)
```
Comment (rsrock) [inline src/QuartzImageIO.jl]: Wouldn't a CFRelease(cg_img) do the trick here?

Also, we can get rid of the semicolons above

File: src/QuartzImageIO.jl (hunk 1 above)
Comment (cstjean) [inline src/QuartzImageIO.jl]: I agree, but isn't the solution with `try` + `finally` even worse? Is there anything cleaner? Or do we simply accept that memory leaks will happen when exceptions are thrown?

File: src/QuartzImageIO.jl (hunk 2 above)
Comment (cstjean) [inline src/QuartzImageIO.jl]: Will do.

**Comment by cstjean** [comment]
I reverted the last two commits and fixed the leak. If you would like, I can squash it as is and we can merge it ready for a tagged release.

As for the error handling code, I'm happy to either leave it to you, or do it myself - if we can agree about what needs to be done. I like the `getenv` example and I can change the current `check_null` calls for that approach. 

For the release of objects, I would use `try` `finally`, but that will also lead to "deep nesting", unless we factor out some of the code into smaller methods. To me that's still better than the current error handling with `if`s (I can see at least two missing `CFRelease(dict)` statements in the current loading code).

Finally, it seems that we should ideally bail out to ImageMagick on every error. I would wrap the whole loading code into something like

```julia
function load(filename)
   try
      load_impl(filename)
   catch
      if debug_flag
          rethrow()
      else
          return nothing # bail out to ImageMagick
      end
   end
end
```

**Comment by timholy** [comment]
Bump. Should I merge this?

**Comment by cstjean** [comment]
I'm fine with merging it. Should I squash first?

**Comment by timholy** [comment]
If you wouldn't mind, that would be great. (No need to squash to a single commit, unless you want to, but at least squashing out the "revert"s would be good.)

**Comment by timholy** [comment]
Also hoping @rsrock or someone can review, but if he's too busy I'll merge by Monday or so.

**Comment by timholy** [comment]
I understand that @rsrock is too busy, so I'll merge whenever you squash.

**Comment by cstjean** [comment]
Done.

**Comment by timholy** [comment]
Thanks!

**Pull request merged**